content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
+it all with system calls just didn't seem like a great use of time. We have some
+functions in the `lib/` directory for printing base types, as well as parsing
+them in order for the `main` function to read command line arguments. A future
+improvement would be putting `malloc` and other C functions we call into this
+file so that we could keep a compiled version around and not have to link with
+C.
+
+## Implementation
+Our interpreter consists of an initial pass to translate into this bytecode,
+which turns temps and labels from strings into numbers, and reduces the size of
+the code overall. The bytecode instructions are stored in an array so that we
+can benefit from cache locality. We then interpret them by iterating through
+this array. One minor inconvenience is that certain instructions, such as print,
+require multiple words. This means that our array index for iteration will
+NOT be an induction variable, so the C compiler will not be able to optimize it
+as aggressively. One solution to this could be using a less flat representation 
+of a list of instructions, and using pointers for the multiword 
+instructions, but this seems like more trouble than it's worth, especially 
+considering the benefits of cache locality.
+
+For the armv8 assembly generation, we initially implemented it as a single pass
+over the bytecode that printed the instructions as it went. This is fine, but it
+leaves little room for extensions and improvements, so in order to get started
+on register allocation we refactored to have an internal representation of arm
+instructions. For this, we went with a tagged union, as it is similar to a
+variant which is very convenient for this, as learned in CS 4120.
+
+## Evaluation
+We used the Bril benchmarks from the repo, and used clock time to evaluate the
+difference. This isn't exactly optimal as most of these benchmarks are very
+small, so the startup time might be significant, but the startup time should
+hurt our interpreter the most, as it has the translation phase. Our results are
+[here](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/eval/performance.ods),
+in ods format. On average, `brili-rs` has an 84% speedup over `brili`, and
+`fastbrili` has a 77% speedup over `brili-rs`. `fastbrili` also has a 96%
+speedup over `brili`, so I think we can call this a job well done.
+
+## Drawbacks
+While we would love to say that `fastbrili` is the answer to all of your Bril
+needs, there are some minor drawbacks, mainly that invalid Bril programs cause
+undefined behavior. Most of the type information is erased and then inferred
+(e.g. if you use `add` on two floats, it will not check, assume they are ints,
+and run integer addition on the values), but we do deal with all correct Bril
+programs correctly. We believe that this isn't much of an issue, as `brili`
+isn't actually that helpful with these errors, though we do acknowledge that
+`fastbrili` is worse.
```

Comment:
I think it's important to reference the semantics for the Bril language, which explicitly permit this kind of behavior (i.e., they are only defined for well-formed programs).